src/box/replication.cc

```diff
@@ -206,6 +210,71 @@ replicaset_set_sync_quorum(void)
 	replication_sync_quorum_auto = replicaset.applier.connected;
 }
 
+/**
+ * Deactivates all expired anonymous replicas. They are expected to be
+ * deleted on deactivation as orphan.
+ * Returns lowest time-to-live among remaining anonymous replicas via
+ * `lowest_ttl`. If there are no anonymous replicas left or the function has
+ * failed to deactivate a replica, `replication_anon_gc_timeout` is returned.
+ */
+static void
+replication_anon_gc(double *lowest_ttl)
+{
+	double timeout = replication_anon_gc_timeout;
+	while (true) {
+		*lowest_ttl = timeout;
+		struct replica *expired_replica = NULL;
+		/*
+		 * Search for the first expired replica. Note that we cannot
+		 * just deactivate it and continue the iteration because a yield
+		 * will happen and the replicaset can be changed significantly.
+		 */
+		replicaset_foreach(replica) {
+			if (!replica->anon || replica_has_connections(replica))
```

Comment:
I am getting lost in the connection states.
* `replica->is_connected`.
* `relay_get_state(replica->relay) == RELAY_FOLLOW`.
* `replica->applier != NULL`.

Interestingly, `replica_has_connections()` doesn't check `replica->is_connected`. Why?

Reply:
> Interestingly, `replica_has_connections()` doesn't check `replica->is_connected`. Why?

Actually, it checks:
```c
bool
replica_has_connections(const struct replica *replica)
{
	assert(replica->relay != NULL);
	/* Relay is expected to be active only for connected replicas. */
	assert(relay_get_state(replica->relay) != RELAY_FOLLOW ||
	       replica->is_connected);
	return replica->is_connected || replica->applier != NULL;
}
```